File: src/engine/MeshDef.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine
{
    public class MeshDef
    {
        public class PointData
        {
            public PointData(int bone, vec3 loc)
            {
                this.boneid = bone;
                this.location = loc;
            }
            public int boneid;
            public vec3 location;

            public override string ToString()
            {
                return string.Format("{0} {1}", boneid, location);
            }
        }

        public List<PointData> points = new List<PointData>();
        public List<vec2> uvs = new List<vec2>();
        public List<vec3> normals = new List<vec3>();
        public List<Bone> bones = new List<Bone>();

        public int TriCount
        {
            get
            {
                int c = 0;
                foreach (KeyValuePair<string, Data> d in datas)
                {
                    c += d.Value.tris.Count;
                }
                return c;
            }
        }

        public MeshDef mapBones()
        {
            Dictionary<int, KeyValuePair<mat44, List<PointData>>> bdp = new Dictionary<int, KeyValuePair<mat44, List<PointData>>>();

            for (int i = 0; i < bones.Count; ++i)
            {
                int parent = bones[i].parent;
                bones[i].index = i;
                if (bones[i].hasParent)
                {
                    bones[i].parentBone = bones[parent];
                    bones[i].parentBone.childs.Add(bones[i]);
                }

                Bone bone = bones[i];
                mat44 m = new MatrixHelper(mat44.Identity).Translate(-bone.pos).Rotate(-bone.rot).mat44;
                bdp.Add(i, new KeyValuePair<mat44, List<PointData>>(m, new List<PointData>()));
            }

            foreach (PointData pd in points)
            {
                if (pd.boneid == -1) continue;
                Bone bone = bones[pd.boneid];
                bdp[pd.boneid].Value.Add(pd);
            }

            foreach(KeyValuePair<int, KeyValuePair<mat44, List<PointData>>> k in bdp )
            {
                Bone bone = bones[k.Key];
                mat44 m = mat44.Identity;
                for(Bone b = bone; b != null; b = b.parentBone)
                {
                    m *= bdp[b.index].Key;
                }

                foreach(PointData pd in k.Value.Value )
                {
                    pd.location = m * pd.location;
                }
            }
            

            return this;
        }

        public IEnumerable<Bone> RootBones
        {
            get
            {
                foreach (Bone b in bones)
                {
                    if (b.parentBone == null)
                        yield return b;
                }
            }
        }

        public class Data
        {
            public List<Tri> tris = new List<Tri>();
        }
        Dictionary<string, Data> datas = new Dictionary<string, Data>();
        Data currentd = null;

        public class MaterialDef 
        {
            public MaterialDef(string name)
            {
                this.name = name;
            }
            public readonly string name;
            public vec3 ambient = new vec3(0.2f, 0.2f, 0.2f);
            public vec3 diffuse = new vec3(0.8f, 0.8f, 0.8f);
            public vec3 specular = new vec3(1,1,1);
            public vec3 emissive = new vec3(0,0,0);

            public float alpha = 1;
            public float shininess = 0;

            public string texture = "";

            public string TextureOrName
            {
                get
                {
                    if (texture == "") return name;
                    else return texture;
                }
            }

            public override string ToString()
            {
                return name + ": " + texture;
            }
        }

        public IEnumerable<Tri> TrianglesFor(MaterialDef material)
        {
            if (datas.ContainsKey(material.name))
            {
                foreach (Tri t in datas[material.name].tris)
                {
                    yield return t;
                }
            }
        }

        Dictionary<string, MaterialDef> materials = new Dictionary<string, MaterialDef>();

        public IEnumerable<MaterialDef> Materials
        {
            get
            {
                foreach (KeyValuePair<string, MaterialDef> k in materials)
                {
                    yield return k.Value;
                }
            }
        }

        public struct VertexData
        {
            public int vertex;
            public int normal;
            public int uv;
        }

        public struct Vertex
        {
            public vec3 vertex;
            public vec3 normal;
            public vec2 uv;
            public int bone;
        }

        public class Bone
        {
            public int parent;
            public string name;
            public vec3 pos;
            public quat rot;

            public int index = 0;
            public Bone parentBone = null;

            public override string ToString()
            {
                if (parentBone == null)
                {
                    return name;
                }
                else return parentBone.ToString() + "->" + name;
            }

            public bool hasParent
            {
                get
                {
                    return parent != -1;
                }
            }

            public List<Bone> childs = new List<Bone>();
        }

        public Vertex[] lookup(Tri tri)
        {
            List<Vertex> v = new List<Vertex>();
            for (int i = 0; i < 3; ++i)
            {
                Vertex c = new Vertex();
                c.vertex = points[tri.data[i].vertex].location;
                c.bone = points[tri.data[i].vertex].boneid;

                if (tri.data[i].normal != -1)
                {
                    c.normal = normals[tri.data[i].normal];
                }

                c.uv = uvs[tri.data[i].uv];
                v.Add(c);
            }
            return v.ToArray();
        }

        public class Tri
        {
            public Tri(VertexData[] data)
            {
                System.Diagnostics.Debug.Assert(data.Length == 3, "It is a triangle");
                this.data = data;
            }
            public VertexData[] data;
        }

        public void addPoint(vec3 p, int bone)
        {
            points.Add( new PointData(bone, p) );
        }

        public void AddUv(vec2 u)
        {
            uvs.Add(u);
        }

        public MaterialDef addMaterial(string name)
        {
            MaterialDef mat = new MaterialDef(name);
            materials.Add(name, mat);
            return mat;
        }

        public void selectMaterial(string name)
        {
            if (materials.ContainsKey(name)==false)
            {
                throw new Exception("mesh does not contain a material named " + name);
            }
            if (datas.ContainsKey(name)==false)
            {
                datas.Add(name, new Data());
            }
            currentd = datas[name];
        }

        public void addTri(Tri t)
        {
            currentd.tris.Add(t);
        }

        public void addNomal(vec3 v)
        {
            normals.Add(v);
        }

        private CompiledMesh compiledMesh = null;

        public CompiledMesh Compiled
        {
            get
            {
                if (compiledMesh == null) throw new Exception("mesh is not compiled");
                return compiledMesh;
            }
        }

        public void compile(MediaLoader ml)
        {
            compiledMesh = new CompiledMesh(ml, this);
        }

        public Bone newBone()
        {
            Bone b = new Bone();
            bones.Add(b);
            return b;
        }

        public bool hasTrianglesFor(MaterialDef m)
        {
            List<Tri> t = new List<Tri>(TrianglesFor(m));
            return t.Count != 0;
        }

        internal void scale(float scale)
        {
            foreach (PointData pd in points)
            {
                pd.location = pd.location * scale;
            }

            foreach (Bone b in bones)
            {
                b.pos = b.pos * scale;
            }
        }

        public MeshDef.MaterialDef getMaterialNamed(string name)
        {
            return materials[name];
        }

        internal void translateFiles(Dictionary<string, string> overrides)
        {
            foreach (MaterialDef d in Materials)
            {
                d.texture = FileSystem.MapFile(overrides, d.TextureOrName);
            }
        }
    }
}

File: src/engine/meshdef.cc
#include "engine/meshdef.h"

#include <cassert>
#include <iostream>
#include <stdexcept>


#include "engine/strings.h"
#include "engine/compiledmesh.h"
#include "engine/filesystem.h"
#include "engine/mat44.h"
#include "engine/matrixhelper.h"
#include "engine/quat.h"
#include "engine/vec2.h"
#include "engine/vec3.h"
#include "fmt/core.h"

namespace simple
{
    SharedVertex::SharedVertex(int b, const vec3& l)
        : bone_id(b)
        , position(l)
    {
    }

    std::string SharedVertex::ToString() const
    {
        return fmt::format("{0} {1}", bone_id, position.ToString());
    }

    std::string Bone::ToString() const
    {
        if (parent_bone == nullptr)
        {
            return name;
        }
        else
            return parent_bone->ToString() + "->" + name;
    }

    bool Bone::HasParent() const
    {
        return parent != -1;
    }

    MaterialDefinition::MaterialDefinition(const std::string& n)
        : name(n)
    {
    }

    std::string MaterialDefinition::GetTextureOrName() const
    {
        return Nullstring(texture, name);
    }

    std::string MaterialDefinition::ToString() const
    {
        return fmt::format("{}: {}", name, Nullstring(texture, "<no_texture>"));
    }

    MatrixAndSharedVertices::MatrixAndSharedVertices()
        : matrix(mat44::Identity())
    {
    }

    MatrixAndSharedVertices::MatrixAndSharedVertices(const mat44& m)
        : matrix(m)
    {
    }

    int MeshDef::GetTriangleCount() const
    {
        int c = 0;
        for (const auto& d : parts)
        {
            c += d.second->size();
        }
        return c;
    }

    MeshDef& MeshDef::UntransformDefaultPose()
    {
        std::map<int, MatrixAndSharedVertices> bone_to_mp;

        // collect all bones and the reverse transformation matrix for the default pose
        for (int i = 0; i < bones.size(); ++i)
        {
            int parent = bones[i]->parent;
            bones[i]->index = i;
            if (bones[i]->HasParent())
            {
                bones[i]->parent_bone = bones[parent];
                bones[i]->parent_bone->children.emplace_back(bones[i]);
            }

            auto& bone = bones[i];
            mat44 m = MatrixHelper(mat44::Identity()).Translate(-bone->position).Rotate(-bone->rotation).AsMat44();
            bone_to_mp.emplace(i, MatrixAndSharedVertices{m});
        }

        // collect all vertices for a each bone... is this really needed? we could easily do this at the end instead
        for (auto& vertex : positions)
        {
            if (vertex->bone_id == -1)
                continue;
            if (vertex->bone_id >= bones.size())
            {
                throw std::runtime_error(fmt::format("bone id {} greater than bones {}", vertex->bone_id, bones.size()));
            }
            auto& mpd = bone_to_mp[vertex->bone_id];
            mpd.shared_vertices.emplace_back(vertex);
        }

        for (auto& k : bone_to_mp)
        {
            const auto current_bone = bones[k.first];

            // for the current bone, combine the local reverse matrices into a total reverse transformation matrix
            auto m = mat44::Identity();
            for (auto loop_bone = current_bone; loop_bone != nullptr; loop_bone = loop_bone->parent_bone)
            {
                m = m * bone_to_mp[loop_bone->index].matrix;
            }

            // finally transform each vertex from the default pose to a "no-pose pose"
            for (const auto& pd : k.second.shared_vertices)
            {
                pd->position = m * pd->position;
            }
        }

        return *this;
    }

    std::vector<std::shared_ptr<Bone>> MeshDef::GetRootBones() const
    {
        std::vector<std::shared_ptr<Bone>> r;

        for (const auto& bone : bones)
        {
            if (bone->parent_bone == nullptr)
            {
                r.emplace_back(bone);
            }
        }

        return r;
    }

    std::vector<Triangle> MeshDef::GetTrianglesFor(const MaterialDefinition& material) const
    {
        const auto found_part = parts.find(material.name);

        if (found_part == parts.end())
        {
            return {};
        }
        else
        {
            return *found_part->second;
        }
    }

    std::vector<std::shared_ptr<MaterialDefinition>> MeshDef::GetMaterials() const
    {
        std::vector<std::shared_ptr<MaterialDefinition>> r;
        for (auto& material_pair : materials)
        {
            r.emplace_back(material_pair.second);
        }
        return r;
    }

    std::vector<Vertex> MeshDef::GetVerticesForTriangle(const Triangle& triangle) const
    {
        std::vector<Vertex> v;
        for (int triangle_index = 0; triangle_index < 3; ++triangle_index)
        {
            Vertex vertex;
            vertex.position = positions[triangle[triangle_index].position]->position;
            vertex.bone = positions[triangle[triangle_index].position]->bone_id;

            if (triangle[triangle_index].normal != -1)
            {
                vertex.normal = normals[triangle[triangle_index].normal];
            }

            vertex.texture_coordinate = texture_coordinates[triangle[triangle_index].texture_coordinate];
            v.emplace_back(vertex);
        }
        return v;
    }

    void MeshDef::AddPosition(const vec3& point, int bone_id)
    {
        positions.emplace_back(std::make_shared<SharedVertex>(bone_id, point));
    }

    void MeshDef::AddTextureCoordinate(const vec2& uv)
    {
        texture_coordinates.emplace_back(uv);
    }

    std::shared_ptr<MaterialDefinition> MeshDef::AddMaterial(const std::string& name)
    {
        auto mat = std::make_shared<MaterialDefinition>(name);
        materials.emplace(name, mat);
        return mat;
    }

    void MeshDef::SelectCurrentMaterial(const std::string& name)
    {
        if (materials.find(name) == materials.end())
        {
            throw std::runtime_error(fmt::format("mesh does not contain a material named {}", name));
        }
        if (parts.find(name) == parts.end())
        {
            parts.emplace(name, std::make_shared<TriangleList>());
        }
        current_part = parts[name];
    }

    void MeshDef::AddTriangle(const Triangle& triangle)
    {
        assert(current_part);
        current_part->emplace_back(triangle);
    }

    void MeshDef::AddNormal(const vec3& v)
    {
        normals.emplace_back(v);
    }

    std::shared_ptr<CompiledMesh> MeshDef::GetCompiledMesh()
    {
        if (compiled_mesh == nullptr)
        {
            assert(false && "mesh is not compiled");
        }
        return compiled_mesh;
    }

    void MeshDef::Compile(MediaLoader* ml)
    {
        compiled_mesh = std::make_shared<CompiledMesh>(ml, this);
    }

    std::shared_ptr<Bone> MeshDef::CreateNewBone()
    {
        auto b = std::make_shared<Bone>();
        bones.emplace_back(b);
        return b;
    }

    bool MeshDef::HasTrianglesFor(MaterialDefinition m)
    {
        auto t = std::vector<Triangle>(GetTrianglesFor(m));
        return t.empty() == false;
    }

    void MeshDef::ScaleMeshAndBones(float scale)
    {
        for (auto& pd : positions)
        {
            pd->position = pd->position * scale;
        }

        for (auto b : bones)
        {
            b->position = b->position * scale;
        }
    }

    std::shared_ptr<MaterialDefinition> MeshDef::GetExistingMaterialNamed(const std::string& name)
    {
        return materials[name];
    }

    void MeshDef::TranslateTexturePaths(const std::map<std::string, std::string>& overrides)
    {
        for (auto d : GetMaterials())
        {
            d->texture = FileSystem::MapFile(overrides, d->GetTextureOrName());
        }
    }
}
